src/test/ci/ratelimiter.test.ts

```diff
@@ -0,0 +1,148 @@
+            });
+        });
+    });
+
+    describe("staggered", () => {
+        const requestsToExpire = 3;
+        const staggerDelay = (queueExpiryTime / (queueCapacity)) * 1000;
+        describe("queue completely filled up, several requests expire, and several more requests enter at the same time", () => {
+            it("should allow all requests to pass, then allow an equal number of expired requests to pass, fail otherwise", async () => {
+                for (let i = 0; i < queueCapacity; i++) {
+                    await delay(staggerDelay * expiryFudgeFactor);
+                    assert.strictEqual(rateLimiter.check(userID), true);
+                }
+
+                assert.strictEqual(rateLimiter.limitMap[userID].length, queueCapacity);
+                await delay((staggerDelay * expiryFudgeFactor) * requestsToExpire);
+
+                for (let i = 0; i < requestsToExpire; i++) {
+                    assert.strictEqual(rateLimiter.check(userID), true);
+                }
+
+                assert.strictEqual(rateLimiter.check(userID), false);
+                assert.strictEqual(rateLimiter.check(userID), false);
+                assert.strictEqual(rateLimiter.check(userID), false);
+                assert.strictEqual(rateLimiter.limitMap[userID].length, queueCapacity);
+            });
+        });
+
+        describe("queue completely filled up, several requests expire, and several more requests enter staggered", () => {
+            it("should allow all requests to pass, then allow an equal number of expired requests to pass, fail otherwise", async () => {
+                for (let i = 0; i < queueCapacity; i++) {
+                    await delay(staggerDelay * expiryFudgeFactor);
+                    assert.strictEqual(rateLimiter.check(userID), true);
+                }
+
+                assert.strictEqual(rateLimiter.limitMap[userID].length, queueCapacity);
+
+                for (let i = 0; i < requestsToExpire; i++) {
+                    await delay(staggerDelay * expiryFudgeFactor);
+                    assert.strictEqual(rateLimiter.check(userID), true);
+                }
+
+                assert.strictEqual(rateLimiter.check(userID), false);
+                assert.strictEqual(rateLimiter.check(userID), false);
+                assert.strictEqual(rateLimiter.check(userID), false);
+                assert.strictEqual(rateLimiter.limitMap[userID].length, queueCapacity);
+            });
+        });
+    });
+});
+
+describe("multiple users", () => {
+    const userID2 = "234";
+    describe("user 1 hits limit, user 2 doesn't", () => {
+        it("should pass all requests until capacity for user 1, and fail else. user 2 should all pass", () => {
+            for (let i = 0; i < queueCapacity; i++) {
+                assert.strictEqual(rateLimiter.check(userID), true);
+            }
+
+            for (let i = 0; i < queueCapacity / 2; i++) {
```

Comment:
Could go up to `queueCapacity - 1` to be a bit more strict with the check

Reply:
This test tests that the requests from one user doesn't impact the requests from another user, intentionally did not fill it to capacity